WebHostLib/options.py

```diff
@@ -17,189 +16,190 @@ def create():
 
     Options.generate_yaml_templates(yaml_folder)
 
-    def get_html_doc(option_type: type(Options.Option)) -> str:
-        if not option_type.__doc__:
-            return "Please document me!"
-        return "\n".join(line.strip() for line in option_type.__doc__.split("\n")).strip()
 
-    weighted_options = {
-        "baseOptions": {
-            "description": "Generated by https://archipelago.gg/",
-            "name": "",
-            "game": {},
+def get_world_theme(game_name: str):
+    if game_name in AutoWorldRegister.world_types:
+        return AutoWorldRegister.world_types[game_name].web.theme
+    return 'grass'
+
+
+def render_options_page(template: str, world_name: str, is_complex: bool = False):
+    world = AutoWorldRegister.world_types[world_name]
+    if world.hidden or world.web.options_page is False:
+        return redirect("games")
+
+    option_groups = {option: option_group.name
+                     for option_group in world.web.option_groups
+                     for option in option_group.options}
+    ordered_groups = ["Game Options", *[group.name for group in world.web.option_groups]]
+    grouped_options = {group: {} for group in ordered_groups}
+    for option_name, option in world.options_dataclass.type_hints.items():
+        # Exclude settings from options pages if their visibility is disabled
+        if not is_complex and option.visibility < Visibility.simple_ui:
+            continue
+
+        if is_complex and option.visibility < Visibility.complex_ui:
+            continue
+
+        grouped_options[option_groups.get(option, "Game Options")][option_name] = option
+
+    return render_template(
+        template,
+        world_name=world_name,
+        world=world,
+        option_groups=grouped_options,
+        issubclass=issubclass,
+        Options=Options,
+        theme=get_world_theme(world_name),
+    )
+
+
+def generate_game(player_name: str, formatted_options: dict):
+    payload = {
+        "race": 0,
+        "hint_cost": 10,
+        "forfeit_mode": "auto",
+        "remaining_mode": "disabled",
+        "collect_mode": "goal",
+        "weights": {
+            player_name: formatted_options,
         },
-        "games": {},
     }
+    r = requests.post("https://archipelago.gg/api/generate", json=payload)
+    if 200 <= r.status_code <= 299:
+        response_data = r.json()
+        return redirect(response_data["url"])
+    else:
+        return r.text
 
-    for game_name, world in AutoWorldRegister.world_types.items():
 
-        all_options: typing.Dict[str, Options.AssembleOptions] = world.options_dataclass.type_hints
+def send_yaml(player_name: str, formatted_options: dict):
+    response = Response(yaml.dump(formatted_options))
```

Comment:
This now sorts the keys of the YAML, as the default for `yaml.dump` is `sort_keys=True`

Personally I would prefer if the YAML maintains the ordering provided by the world and presented on the page. Worlds generally group related options together, so maintaining that in the YAML would improve the experience of manually editing it for players and reviewing it for hosts and async helpers. 

```suggestion
    response = Response(yaml.dump(formatted_options, sort_keys=False))
```

Reply:
That's extremely helpful. I was wondering if there was a way to prevent the manual re-ordering. Change pushed.